csharp/ql/lib/semmle/code/csharp/security/dataflow/ZipSlipQuery.qll

```diff
@@ -3,147 +3,360 @@
+  PathCombinerViaStringConcatenation() { exists(AddExpr e | this = e) }
+}
 
-  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+class MethodCallGetFullPath extends MethodCall {
+  MethodCallGetFullPath() { this.getTarget().hasQualifiedName("System.IO.Path", "GetFullPath") }
+}
 
-  override predicate isSanitizer(DataFlow::Node node) { node instanceof Sanitizer }
+class GetFullPathToQualifierTaintTrackingConfiguration extends TaintTracking4::Configuration {
+  GetFullPathToQualifierTaintTrackingConfiguration() {
+    this = "GetFullPathToQualifierTaintTrackingConfiguration"
+  }
+
+  override predicate isSource(DataFlow::Node node) {
+    exists(MethodCallGetFullPath mcGetFullPath | node = DataFlow::exprNode(mcGetFullPath))
+  }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodCall mc |
+      mc.getTarget() instanceof RootSanitizerMethod and
+      node.asExpr() = mc.getQualifier()
+    )
+  }
+}
+
+class PathCombinerToGetFullPathTaintTrackingConfiguration extends TaintTracking4::Configuration {
+  PathCombinerToGetFullPathTaintTrackingConfiguration() {
+    this = "PathCombinerToGetFullPathTaintTrackingConfiguration "
+  }
+
+  override predicate isSource(DataFlow::Node node) {
+    exists(UnsanitizedPathCombiner pathCombiner | node = DataFlow::exprNode(pathCombiner))
+  }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodCallGetFullPath mcGetFullPath |
+      node = DataFlow::exprNode(mcGetFullPath.getArgument(0))
+    )
+  }
+}
+
+private predicate safeCombineGetFullPathSequence(MethodCallGetFullPath mcGetFullPath, Expr q) {
+  // Path.Combine THEN Path.GetFullPath is applied (with possibly arbitrary mutations)
+  exists(DataFlow::Node source, PathCombinerToGetFullPathTaintTrackingConfiguration taintTracking |
+    taintTracking.hasFlow(source, DataFlow::exprNode(mcGetFullPath.getArgument(0)))
+  ) and
+  exists(GetFullPathToQualifierTaintTrackingConfiguration qualifierTaintTracker |
+    qualifierTaintTracker.hasFlow(DataFlow::exprNode(mcGetFullPath), DataFlow::exprNode(q))
+  )
+}
+
+class RootSanitizerMethodCall extends SanitizerMethodCall {
+  /*
+   * The set of /valid/ Guards of RootSanitizerMethodCall.
+   *
+   *    IN CONJUNCTION with BOTH
+   *      Path.Combine
+   *      AND Path.GetFullPath
+   *    OR
+   *      There is a direct flow from Path.GetFullPath to qualifier of RootSanitizerMethodCall.
+   *
+   *    It is not simply enough for the qualifier of String.StartsWith
+   *    to pass through Path.Combine without also passing through GetFullPath AFTER.
+   */
+
+  RootSanitizerMethodCall() {
+    exists(RootSanitizerMethod sm | this.getTarget() = sm) and
+    exists(Expr q, AbstractValue v |
+      this.getQualifier() = q and
+      v.(AbstractValues::BooleanValue).getValue() = true and
+      exists(MethodCallGetFullPath mcGetFullPath | safeCombineGetFullPathSequence(mcGetFullPath, q))
```

Comment:
There are a lot of references to a class that is simply an alias for a certain method. Just a suggestion but maybe we have a naming convention so it is easier to understand at a glance? for example, "System.IO.Path.GetFullPath" can have a codeql class called SystemIOPathGetFullPath. we already do this in some of our libraries actually. that would make chunks of code like 99-104 much more readable

Reply:
Updated RootSanitizerMethod to be `MethodSystemStringStartsWith`. I will keep `RootSanitizerMethodCall` as-is however because the stack abstraction here is more important to convey than the implementing method.